File: FuncScript/Functions/Text/RegexFunction.cs
using FuncScript.Core;
using FuncScript.Model;
using System.Text.RegularExpressions;

namespace FuncScript.Functions.Text
{
    [ProviderCollection("text")]
    public class RegexFunction : IFsFunction
    {
        public int MaxParsCount => 3;

        public CallType CallType => CallType.Prefix;

        public string Symbol => "regex";

        public int Precedence => 0;

        public object Evaluate(object par)
        {
            var pars = FunctionArgumentHelper.ExpectList(par, this.Symbol);

            if (pars.Length < 2 || pars.Length > MaxParsCount)
                return new FsError(FsError.ERROR_PARAMETER_COUNT_MISMATCH, $"{this.Symbol}: two or three parameters expected");

            var textValue = pars[0];
            var patternValue = pars[1];
            var flagsValue = pars.Length > 2 ? pars[2] : null;

            if (textValue == null || patternValue == null)
                return new FsError(FsError.ERROR_TYPE_MISMATCH, $"{this.Symbol}: text and pattern are required");

            if (textValue is not string text)
                return new FsError(FsError.ERROR_TYPE_MISMATCH, $"{this.Symbol}: text parameter must be string");

            if (patternValue is not string pattern)
                return new FsError(FsError.ERROR_TYPE_MISMATCH, $"{this.Symbol}: pattern parameter must be string");

            if (!TryParseOptions(flagsValue, out var options, out var optionError))
                return optionError;

            return Regex.IsMatch(text, pattern, options);
        }

        static bool TryParseOptions(object flagsValue, out RegexOptions options, out FsError error)
        {
            options = RegexOptions.CultureInvariant;
            error = null;

            if (flagsValue == null)
                return true;

            if (flagsValue is string flagsText)
            {
                foreach (var ch in flagsText)
                {
                    if (char.IsWhiteSpace(ch) || ch == ',' || ch == '|')
                        continue;

                    switch (char.ToLowerInvariant(ch))
                    {
                        case 'i':
                            options |= RegexOptions.IgnoreCase;
                            break;
                        case 'm':
                            options |= RegexOptions.Multiline;
                            break;
                        case 's':
                            options |= RegexOptions.Singleline;
                            break;
                        case 'x':
                            options |= RegexOptions.IgnorePatternWhitespace;
                            break;
                        default:
                            error = new FsError(FsError.ERROR_TYPE_INVALID_PARAMETER, $"regex: unsupported regex option '{ch}'");
                            return false;
                    }
                }

                return true;
            }

            error = new FsError(FsError.ERROR_TYPE_MISMATCH, "regex: flags parameter must be string");
            return false;
        }

        public string ParName(int index)
        {
            return index switch
            {
                0 => "text",
                1 => "pattern",
                2 => "flags",
                _ => string.Empty
            };
        }
    }
}

File: js-port/funcscript-js/src/funcs/text/regex-function.js
const { BaseFunction, CallType } = require('../../core/function-base');
const helpers = require('../helpers');
const { FSDataType } = require('../../core/fstypes');

class RegexFunction extends BaseFunction {
  constructor() {
    super();
    this.symbol = 'regex';
    this.callType = CallType.Prefix;
  }

  get maxParameters() {
    return 3;
  }

  evaluate(provider, parameters) {
    if (parameters.count < 2 || parameters.count > this.maxParameters) {
      return helpers.makeError(helpers.FsError.ERROR_PARAMETER_COUNT_MISMATCH, `${this.symbol}: two or three parameters expected`);
    }

    const textResult = helpers.requireString(this.symbol, parameters.getParameter(provider, 0), 'text');
    if (!textResult.ok) {
      return textResult.error;
    }
    const text = textResult.value;

    const patternResult = helpers.requireString(this.symbol, parameters.getParameter(provider, 1), 'pattern');
    if (!patternResult.ok) {
      return patternResult.error;
    }
    let pattern = patternResult.value;

    let flagsText = null;
    if (parameters.count === 3) {
      const flagValue = helpers.assertTyped(parameters.getParameter(provider, 2));
      if (helpers.typeOf(flagValue) !== FSDataType.Null) {
        if (helpers.typeOf(flagValue) !== FSDataType.String) {
          return helpers.makeError(helpers.FsError.ERROR_TYPE_MISMATCH, `${this.symbol}: flags parameter must be a string`);
        }
        flagsText = helpers.valueOf(flagValue);
      }
    }

    const flagResult = this.parseFlags(flagsText || '');
    if (!flagResult.ok) {
      return flagResult.error;
    }

    if (flagResult.ignoreWhitespace) {
      pattern = RegexFunction.stripPatternWhitespace(pattern);
    }

    try {
      const regex = new RegExp(pattern, flagResult.flags);
      return helpers.makeValue(FSDataType.Boolean, regex.test(text));
    } catch (error) {
      return helpers.makeError(helpers.FsError.ERROR_TYPE_INVALID_PARAMETER, `${this.symbol}: invalid regular expression`);
    }
  }

  parseFlags(flagsText) {
    if (!flagsText) {
      return { ok: true, flags: '', ignoreWhitespace: false };
    }

    let ignoreWhitespace = false;
    let flags = '';
    const seen = new Set();
    for (const raw of flagsText) {
      if (raw === ',' || raw === '|') {
        continue;
      }
      if (/\s/.test(raw)) {
        continue;
      }
      const ch = raw.toLowerCase();
      if (seen.has(ch)) {
        continue;
      }
      seen.add(ch);
      switch (ch) {
        case 'i':
          flags += 'i';
          break;
        case 'm':
          flags += 'm';
          break;
        case 's':
          flags += 's';
          break;
        case 'x':
          ignoreWhitespace = true;
          break;
        default:
          return {
            ok: false,
            error: helpers.makeError(helpers.FsError.ERROR_TYPE_MISMATCH, `${this.symbol}: unsupported regex option '${raw}'`)
          };
      }
    }

    return { ok: true, flags, ignoreWhitespace };
  }

  static stripPatternWhitespace(pattern) {
    let result = '';
    let escaping = false;
    let inCharClass = false;
    for (let i = 0; i < pattern.length; i += 1) {
      const ch = pattern[i];
      if (escaping) {
        result += ch;
        escaping = false;
        continue;
      }
      if (ch === '\\') {
        escaping = true;
        result += ch;
        continue;
      }
      if (ch === '[') {
        inCharClass = true;
        result += ch;
        continue;
      }
      if (ch === ']' && inCharClass) {
        inCharClass = false;
        result += ch;
        continue;
      }
      if (!inCharClass && ch === '#') {
        while (i + 1 < pattern.length) {
          const next = pattern[i + 1];
          if (next === '\\n' || next === '\\r') {
            break;
          }
          i += 1;
        }
        continue;
      }
      if (!inCharClass && /\s/.test(ch)) {
        continue;
      }
      result += ch;
    }
    if (escaping) {
      result += '\\';
    }
    return result;
  }

  parName(index) {
    return index === 0 ? 'text' : index === 1 ? 'pattern' : index === 2 ? 'flags' : '';
  }
}

module.exports = {
  RegexFunction
};
